lib/os/sem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	if (sem == NULL || limit == 0 ||
+	    initial_count > limit || limit > INT_MAX) {
+		return -EINVAL;
+	}
+
+	atomic_set(&sem->futex.val, initial_count);
+	sem->limit = limit;
+
+	return 0;
+}
+
+unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	int value = atomic_get(&sem->futex.val);
+
+	return value > 0 ? value : 0;
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
```

Comment:
Style: you're using this pattern repeatedly, and it's actually fairly hard to understand.  How about abstract out something like:

```c
static inline atomic_t bounded_dec(atomic_t *val, atomic_t minimum)
{
    atomic_t old_value, new_value;

    do {
           old_value = atomic_get(val);
	   new_value = val <= minimum ? minimum : val - 1;
    } while(atomic_cas(val, old_value, new_value) == 0);

    return old_value;
}

static inline atomic_t bounded_inc(atomic_t *val, atomic_t maximum)
{
    atomic_t old_value, new_value;

    do {
           old_value = atomic_get(val);
	   new_value = val >= maximum ? maximum : val + 1;
    } while(atomic_cas(val, old_value, new_value) == 0);

    return old_value;
}
```

IMHO, this would clarify your atomic loops a lot, while generating the same code.

Reply:
thanks, I will do the abstract as you suggested.

Reply:
@wentongwu please address this

Reply:
@andrewboie  update the code , please take a look.